File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {
    const N = words.length;
    const visited = Array.from({length: N}, ()=>false);
    const queue = [[begin, 0]];
    
    let head = 0;
    while(head<queue.length){
        const [word, count] = queue[head++];
        if(word===target)   return count;
        
        for(let i=0; i < N; i++){
            if(!visited[i] && isValid(word, words[i])){
                visited[i] = true;
                queue.push([words[i], count + 1])
            }
        }
    }
    
    
    return 0;
}

function isValid(A, B){
    let diff = 0;
    for(let i=0; i<A.length; i++){
        if(A[i]!==B[i]) diff++;
    }
    
    return diff === 1;
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def solution(begin, target, words):
    answer = 0
    N = len(words)
    
    visited = [False] * N
    queue = deque()
    queue.append((begin, 0))
    
    while queue:
        word, cnt = queue.popleft()
        
        if word == target:
            return cnt
        
        for i, next_word in enumerate(words):
            if not visited[i] and isOneDiff(word, next_word):
                visited[i] = True
                queue.append((next_word, cnt+1))
    
    
    return 0

def isOneDiff(a, b):
    diff = 0
    
    for i in range(len(a)):
        if diff > 1:
            return False
        if a[i] != b[i]:
            diff+=1
            
    return diff == 1